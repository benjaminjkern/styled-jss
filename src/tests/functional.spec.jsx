import 'react-dom'
import React from 'react'
import {mount} from 'enzyme'

import {
  getCss,
  removeWhitespace
} from './utils'

import CreateApp from './App'

let Styled
let styled

const mockNameGenerators = () => {
  let styledCounter = 0

  jest.mock('../utils/generateTagName')
  jest.mock('jss/lib/utils/generateClassName')

  const generateTagName = require('../utils/generateTagName').default
  const generateClassName = require('jss/lib/utils/generateClassName').default

  // $FlowIgnore
  generateTagName.mockImplementation((tagName: string) => `${tagName}-${++styledCounter}`)
  generateClassName.mockImplementation(rule => `${rule.name}-id`)
}

const assertSheet = (sheet) => {
  expect(sheet.toString()).toMatchSnapshot()
  expect(getCss(sheet)).toBe(removeWhitespace(sheet.toString()))
}

describe('functional tests', () => {
  beforeEach(() => {
    mockNameGenerators()

    Styled = require('../').Styled
    styled = Styled()
  })

  afterEach(() => {
    styled.mountSheet().detach()
  })

  it('should update props and unmount', () => {
    const App = CreateApp(styled)
    const wrapper = mount(<App />)
    const sheet = styled.mountSheet()

    assertSheet(sheet)
    wrapper.setProps({margin: 20})
    assertSheet(sheet)
    wrapper.unmount()
  })

  it('should update nested props', () => {
    styled = Styled({
      button: {
        fontSize: 12,
      },
    })

    const Div = styled('div')({
      padding: 15,
      '&:hover': {
        '& $button': {
          color: ({primary}) => (primary ? 'red' : 'green'),
        },
      },
    })

    const Button = styled('button')({
      composes: '$button',
    })

    const App = (props: {primary?: boolean}) => (
      <div>
        <Div primary={props.primary}>
          <Button>Button</Button>
        </Div>
        <Div primary={!props.primary}>
          <Button>Button</Button>
        </Div>
      </div>
    )

    const wrapper = mount(<App />)
    const sheet = styled.mountSheet()

    assertSheet(sheet)
    wrapper.setProps({primary: true})
    assertSheet(sheet)
    wrapper.unmount()
  })
})
