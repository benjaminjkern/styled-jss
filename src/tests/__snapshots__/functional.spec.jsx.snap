// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`functional tests should update dynamic props for conditional rules 1`] = `
".button-1-id {
  padding: 10px;
}
@media screen {
  .button-1-id .button {
    margin: 10px;
  }
}"
`;

exports[`functional tests should update dynamic props for conditional rules 2`] = `
".button-1-id {
  padding: 0;
}
@media screen {
  .button-1-id .button {
    margin: 0;
  }
}"
`;

exports[`functional tests should update nested props 1`] = `
".button-id {
  font-size: 12px;
}
.div-1-id {
  padding: 15px;
}
.div-3-id:hover .button-id {
  color: green;
}
.div-4-id:hover .button-id {
  color: red;
}"
`;

exports[`functional tests should update nested props 2`] = `
".button-id {
  font-size: 12px;
}
.div-1-id {
  padding: 15px;
}
.div-3-id:hover .button-id {
  color: red;
}
.div-4-id:hover .button-id {
  color: green;
}"
`;

exports[`functional tests should update props and unmount 1`] = `
".div-1-id {
  margin: 50px;
}
.header-2-id {
  padding: 10px;
}
.h1-5-id {
  color: red;
}
.section-3-id {
  color: red;
}
.button-6-id {
  margin: 0;
}
.button-7-id {
  margin: 10px;
}
.section-4-id {
  color: yellow;
}"
`;

exports[`functional tests should update props and unmount 2`] = `
".div-1-id {
  margin: 50px;
}
.header-2-id {
  padding: 10px;
}
.h1-5-id {
  color: red;
}
.section-3-id {
  color: red;
}
.button-6-id {
  margin: 0;
}
.button-7-id {
  margin: 20px;
}
.section-4-id {
  color: yellow;
}"
`;

exports[`functional tests should use props on remount 1`] = `
".button-1-id {
  color: black;
}"
`;

exports[`functional tests should use props on remount 2`] = `
".button-1-id {
  color: red;
}"
`;
